File: backend_app/app/utils/distance.py
"""Distance calculation utilities."""
import math
from typing import List


def calculate_haversine_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """
    Calculate distance between two coordinates using Haversine formula (km).
    
    Args:
        lat1: Latitude of first point
        lng1: Longitude of first point
        lat2: Latitude of second point
        lng2: Longitude of second point
    
    Returns:
        Distance in kilometers
    """
    R = 6371  # Earth radius in km
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lng = math.radians(lng2 - lng1)
    
    a = (
        math.sin(delta_lat / 2) ** 2 +
        math.cos(lat1_rad) * math.cos(lat2_rad) * math.sin(delta_lng / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return R * c


def calculate_total_distance(steps: List) -> float:
    """
    Calculate total distance traveled across all steps.
    
    Args:
        steps: List of step objects with lat/lng attributes
    
    Returns:
        Total distance in kilometers
    """
    if not steps or len(steps) < 2:
        return 0.0
    
    total_distance = 0.0
    for i in range(len(steps) - 1):
        distance = calculate_haversine_distance(
            steps[i].lat,
            steps[i].lng,
            steps[i + 1].lat,
            steps[i + 1].lng
        )
        total_distance += distance
    
    return round(total_distance, 2)

File: frontend/lib/distance.ts
import { Step } from "./types";

/**
 * Calculate distance between two geographic coordinates using Haversine formula.
 * @param lat1 - Latitude of first point in degrees
 * @param lon1 - Longitude of first point in degrees
 * @param lat2 - Latitude of second point in degrees
 * @param lon2 - Longitude of second point in degrees
 * @returns Distance in kilometers
 */
export function calculateDistance(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371; // Earth's radius in kilometers
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(R * c * 100) / 100; // Round to 2 decimal places
}

/**
 * Calculate total distance traveled across all steps.
 * @param steps - Array of step objects with lat/lng properties
 * @returns Total distance in kilometers
 */
export function calculateTotalDistance(steps: Step[]): number {
  if (!steps || steps.length < 2) return 0;

  let totalDistance = 0;
  for (let i = 0; i < steps.length - 1; i++) {
    const distance = calculateDistance(
      steps[i].lat,
      steps[i].lng,
      steps[i + 1].lat,
      steps[i + 1].lng
    );
    totalDistance += distance;
  }

  return Math.round(totalDistance * 100) / 100; // Round to 2 decimal places
}

/**
 * Get distances for each segment between consecutive steps.
 * @param steps - Array of step objects with lat/lng properties
 * @returns Array of distances for each segment
 */
export function calculateSegmentDistances(steps: Step[]): number[] {
  if (!steps || steps.length < 2) return [];

  const distances: number[] = [];
  for (let i = 0; i < steps.length - 1; i++) {
    const distance = calculateDistance(
      steps[i].lat,
      steps[i].lng,
      steps[i + 1].lat,
      steps[i + 1].lng
    );
    distances.push(distance);
  }

  return distances;
}
